orchagent/pfcwdorch.cpp

```diff
@@ -0,0 +1,733 @@
+        return;
+    }
+
+    if (port.m_type != Port::PHY)
+    {
+        SWSS_LOG_ERROR("Interface %s is not physical port", key.c_str());
+        return;
+    }
+
+    for (auto i : data)
+    {
+        const auto &field = fvField(i);
+        const auto &value = fvValue(i);
+
+        try
+        {
+            if (field == PFC_WD_DETECTION_TIME)
+            {
+                detectionTime = to_uint<uint32_t>(
+                        value,
+                        PFC_WD_DETECTION_TIME_MIN,
+                        PFC_WD_DETECTION_TIME_MAX);
+            }
+            else if (field == PFC_WD_RESTORATION_TIME)
+            {
+                restorationTime = to_uint<uint32_t>(value,
+                        PFC_WD_RESTORATION_TIME_MIN,
+                        PFC_WD_RESTORATION_TIME_MAX);
+            }
+            else if (field == PFC_WD_ACTION)
+            {
+                action = deserializeAction(value);
+                if (action == PfcWdAction::PFC_WD_ACTION_UNKNOWN)
+                {
+                    SWSS_LOG_ERROR("Invalid PFC Watchdog action %s", value.c_str());
+                    return;
+                }
+            }
+            else
+            {
+                SWSS_LOG_ERROR(
+                        "Failed to parse PFC Watchdog %s configuration. Unknown attribute %s.\n",
+                        key.c_str(),
+                        field.c_str());
+                return;
+            }
+        }
+        catch (const exception& e)
+        {
+            SWSS_LOG_ERROR(
+                    "Failed to parse PFC Watchdog %s attribute %s error: %s.",
+                    key.c_str(),
+                    field.c_str(),
+                    e.what());
+            return;
+        }
+        catch (...)
+        {
+            SWSS_LOG_ERROR(
+                    "Failed to parse PFC Watchdog %s attribute %s. Unknown error has been occurred",
+                    key.c_str(),
+                    field.c_str());
+            return;
+        }
+    }
+
+    // Validation
+    if (detectionTime == 0)
+    {
+        SWSS_LOG_ERROR("%s missing", PFC_WD_DETECTION_TIME);
+        return;
+    }
+    
+    if (restorationTime == 0)
+    {
+        SWSS_LOG_ERROR("%s missing", PFC_WD_RESTORATION_TIME);
+        return;
+    }
+
+    // Start Watchdog on every lossless queue
+    sai_attribute_t attr;
+    attr.id = SAI_PORT_ATTR_PRIORITY_FLOW_CONTROL;
+
+    sai_status_t status = sai_port_api->get_port_attribute(port.m_port_id, 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get PFC mask on port %s: %d", port.m_alias.c_str(), status);
+        return;
+    }
+
+    uint8_t pfcMask = attr.value.u8;
+    for (uint8_t i = 0; i < PFC_WD_TC_MAX; i++)
+    {
+        if ((pfcMask & (1 << i)) == 0)
+        {
+            continue;
+        }
+
+        sai_object_id_t queueId = port.m_queue_ids[i];
+
+        initWdCounters(sai_serialize_object_id(queueId));
```

Comment:
Use lock?

Reply:
WD thread is not yet running on that queue, so we can update counters without worrying about race condition,